File: JavaScript/DSA/Stack.js
// A Stack is a linear data structure that follows the Last-In-First-Out (LIFO) principle, which means that the last element added to the stack will be the first one to be removed

class Stack {
  constructor() {
    this.items = [];
  }

  // add an item to the top of the stack
  push(item) {
    this.items.push(item);
  }

  // remove and return the top item from the stack
  pop() {
    if (this.isEmpty()) {
      return null;
    }
    return this.items.pop();
  }

  // return the top item from the stack without removing it
  peek() {
    if (this.isEmpty()) {
      return null;
    }
    return this.items[this.items.length - 1];
  }

  // return true if the stack is empty, false otherwise
  isEmpty() {
    return this.items.length === 0;
  }

  // return the number of items in the stack
  size() {
    return this.items.length;
  }

  // print the contents of the stack
  print() {
    console.log(this.items);
  }
}

// usage

const myStack = new Stack();

myStack.push(10);
myStack.push(20);
myStack.push(30);
myStack.print(); // Output: [10, 20, 30]

console.log(myStack.pop()); // Output: 30
console.log(myStack.peek()); // Output: 20

myStack.print(); // Output: [10, 20]
console.log(myStack.size()); // Output: 2
console.log(myStack.isEmpty()); // Output: false

File: C&C++/DSA/Stack.cpp
/*
A stack in C++ is a linear data structure that follows the Last-In-First-Out (LIFO) principle. It means that the last element inserted into the stack is the first one to be removed. Stacks can be implemented using arrays
*/

#include <iostream>

using namespace std;

const int MAX_SIZE = 100;

class Stack
{
private:
    int arr[MAX_SIZE];
    int top;

public:
    Stack()
    {
        top = -1;
    }

    bool is_empty()
    {
        return top == -1;
    }

    bool is_full()
    {
        return top == MAX_SIZE - 1;
    }

    void push(int value)
    {
        if (is_full())
        {
            cout << "Stack overflow!" << endl;
            return;
        }

        arr[++top] = value;
    }

    void pop()
    {
        if (is_empty())
        {
            cout << "Stack underflow!" << endl;
            return;
        }

        top--;
    }

    int peek()
    {
        if (is_empty())
        {
            cout << "Stack is empty!" << endl;
            return -1;
        }

        return arr[top];
    }

    void print()
    {
        if (is_empty())
        {
            cout << "Stack is empty!" << endl;
            return;
        }

        for (int i = top; i >= 0; i--)
        {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};

int main()
{
    Stack stack;

    stack.push(1);
    stack.push(2);
    stack.push(3);

    stack.print(); // Output: 3 2 1

    stack.pop();

    stack.print(); // Output: 2 1

    cout << "Top element: " << stack.peek() << endl; // Output: Top element: 2

    stack.pop();
    stack.pop();
    stack.pop(); // Output: Stack underflow!

    return 0;
}
